Example concert ticket using template

kellemNegasi opened the pull request:
Creates concert ticket using templates according to https://unidoc.atlassian.net/browse/UP-407

sampila commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                                    <text-chunk font="helvetica" font-size="10">E - ticket</text-chunk>
+                                </paragraph>
+                                <paragraph text-align="left" margin="2 0 0 10">
+                                    <text-chunk font="helvetica" font-size="10">000385724</text-chunk>
```
Probably can make this value as parameter?

sampila commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                    <division>
+                        <image src="qr-code" height="75" width="75" margin="15 0 0 10"></image>
+                        <paragraph text-align="top" margin="0 0 0 15">
+                            <text-chunk font="helvetica" font-size="10">000385724</text-chunk>
```
This also, as this value can be part of qrcode?

sampila reviewed (commented):
Great works, added minor comments.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+    <table-cell vertical-align="top">
+        <image src="path('./templates/res/red-guitar.png')" width="100" height="160" margin="5 0 0 0"></image>
+    </table-cell>
+    <table-cell vertical-align="top">
```
`top` is the default for the `vertical-align` attribute. No need to specify it.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            <division margin="0 0 0 15">
+            {{template "header" dict "FontSize" 19 "SubFont" 11}}
+            </division>
+                <table columns="2" padding="0, 0, 0, 0" column-widths="0.3 0.7">
```
No need to specify the padding. `0` is the default value. Also, the syntax of the attribute is not correct. The values should be separated by spaces.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                    <table columns="1" margin="10 0 0 0">
+                        <table-cell>
+                            <division margin="0 0 0 0" padding="0, 0, 0, 0">
+                                <paragraph text-align="left" margin="0 0 0 10" line-height="1.1">
```
`left` is the default for `text-align`. No need to specify the attribute.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                        <table-cell>
+                            <division margin="0 0 0 0" padding="0, 0, 0, 0">
+                                <paragraph text-align="left" margin="0 0 0 10" line-height="1.1">
+                                    <text-chunk font="helvetica" font-size="10">E - ticket</text-chunk>
```
`helvetica` is the default font for text chunks. Also, `font-size` is 10 by default. No need to specify these attributes.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                                <paragraph text-align="left" margin="0 0 0 10" line-height="1.1">
+                                    <text-chunk font="helvetica" font-size="10">E - ticket</text-chunk>
+                                </paragraph>
+                                <paragraph text-align="left" margin="2 0 0 10">
```
`left` is the default for `text-align`. No need to specify the attribute.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                                    <text-chunk font="helvetica" font-size="10">E - ticket</text-chunk>
+                                </paragraph>
+                                <paragraph text-align="left" margin="2 0 0 10">
+                                    <text-chunk font="helvetica" font-size="10">{{.TicketNumber}}</text-chunk>
```
`helvetica` is the default font for text chunks. Also, `font-size` is 10 by default. No need to specify these attributes.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+</table>
+<table columns="2" margin="0 0 0 15">
+ <table-cell border-width-top="0.5" border-width-bottom="0.5" border-width-left="0.5" vertical-align="top" indent="0.5" border-style="single">
+    <division margin="0 0 15 15" padding="0, 0, 0, 0">
```
No need to specify the padding. `0` is the default value. Also, the syntax of the attribute is not correct. The values should be separated by spaces.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            <text-chunk font="helvetica-bold" font-size="12">Rules of attendance</text-chunk>
+        </paragraph>
+        {{range .RulesOfAttendance}}
+            <paragraph text-align="left" margin="0 0 0 0" line-height="1.3">
```
No need to specify the margin. `0` is the default value.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+        </paragraph>
+        {{range .RulesOfAttendance}}
+            <paragraph text-align="left" margin="0 0 0 0" line-height="1.3">
+                <text-chunk font="helvetica" font-size="9">{{.}}</text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+        {{end}}
+    </division>
+ </table-cell>
+ <table-cell border-width-top="0.5" border-width-bottom="0.5" border-width-right="0.5" vertical-align="top" indent="0" border-style="single">
```
`top` is the default for the `vertical-align` attribute. No need to specify it.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+    </division>
+ </table-cell>
+ <table-cell border-width-top="0.5" border-width-bottom="0.5" border-width-right="0.5" vertical-align="top" indent="0" border-style="single">
+    <division margin="0 0 15 15" padding="0, 0, 0, 0">
```
No need to specify the padding. `0` is the default value. Also, the syntax of the attribute is not correct. The values should be separated by spaces.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            <text-chunk font="helvetica-bold" font-size="12">Rules of purchase</text-chunk>
+        </paragraph>
+        {{range .RulesOfPurchase}}
+            <paragraph text-align="left" margin="0 0 0 0" line-height="1.3">
```
No need to specify the margin. `0` is the default value.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+        </paragraph>
+        {{range .RulesOfPurchase}}
+            <paragraph text-align="left" margin="0 0 0 0" line-height="1.3">
+                <text-chunk font="helvetica" font-size="9">{{.}}</text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+
+<table columns="3" margin="10 0 0 10" column-widths="0.5 0.20 0.30">
+    <table-cell vertical-align="top">
+        <division margin="0 20 0 0" padding="0, 0, 0, 0">
```
No need to specify the padding. `0` is the default value. Also, the syntax of the attribute is not correct. The values should be separated by spaces.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+<table columns="3" margin="10 0 0 10" column-widths="0.5 0.20 0.30">
+    <table-cell vertical-align="top">
+        <division margin="0 20 0 0" padding="0, 0, 0, 0">
+            <paragraph text-align="top" margin="0 0 0 0">
```
No need to specify the margin. `0` is the default value.
Also, `top` is not a valid value for the `text-align` attribute of paragraphs. Valid values are `left`, `right`, `center`, `justify`.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                <text-chunk font="helvetica-bold" font-size="12">Program/Bands List/Event Description</text-chunk>
+            </paragraph>
+            <paragraph text-align="top" margin="5 0 0 0" line-height="1.3">
+                <text-chunk font="helvetica" font-size="9">A City literally built on Rock n' Roll in Cadott WI, </text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            </paragraph>
+        </division>
+    </table-cell>
+    <table-cell vertical-align="top">
```
`top` is the default for the `vertical-align` attribute. No need to specify it.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+    <table-cell vertical-align="top">
+        <image src="path('./templates/res/map.png')" height="160" width="100" margin="5 0 0 0"></image>
+    </table-cell>
+    <table-cell vertical-align="top">
```
`top` is the default for the `vertical-align` attribute. No need to specify it.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            <paragraph text-align="top" margin="0 0 0 0">
+                <text-chunk font="helvetica-bold" font-size="12">Program/Bands List/Event Description</text-chunk>
+            </paragraph>
+            <paragraph text-align="top" margin="5 0 0 0" line-height="1.3">
```
`top` is not a valid value for the `text-align` attribute of paragraphs.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+    </table-cell>
+    <table-cell vertical-align="top">
+        <division>
+            <paragraph text-align="top" margin="0 0 0 0">
```
`top` is not a valid value for the `text-align` attribute of paragraphs.
Also, the default value for the `margin` attribute is `0`. No need to specify it in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                    <text-chunk font="helvetica-bold" font-size="12">How to find us</text-chunk>
+            </paragraph>
+            <paragraph text-align="top" margin="5 0 0 0" line-height="1.3">
+                <text-chunk font="helvetica" font-size="9">Concert Hall is located three blocks west of the park and three blocks north the library.</text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+            <paragraph text-align="top" margin="0 0 0 0">
+                    <text-chunk font="helvetica-bold" font-size="12">How to find us</text-chunk>
+            </paragraph>
+            <paragraph text-align="top" margin="5 0 0 0" line-height="1.3">
```
`top` is not a valid value for the `text-align` attribute of paragraphs.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                <text-chunk font="helvetica" font-size="9">Free parking is available nearby, which fills up close to showtime, and on the streets around.</text-chunk>
+                <text-chunk font="helvetica" font-size="9">Bike racks are located outside the main entrance tothe Hall.</text-chunk>
+            </paragraph>
+            <paragraph text-align="top" margin="10 0 0 0">
```
`top` is not a valid value for the `text-align` attribute of paragraphs.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                <text-chunk font="helvetica-bold" font-size="12">Learn More At</text-chunk>
+            </paragraph>
+            <paragraph margin="5 0 0 0">
+                <text-chunk font="helvetica" font-size="9">Facebook /ConcertHall</text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+<table columns="2">
+    <table-cell vertical-align="middle">
+        <division margin="0 0 0 10">
+            <paragraph text-align="top" margin="0 0 0 0">
```
`top` is not a valid value for the `text-align` attribute of paragraphs.
Also, the default value for the `margin` attribute is `0`. No need to specify it in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+    <table-cell vertical-align="middle">
+        <division margin="0 0 0 10">
+            <paragraph text-align="top" margin="0 0 0 0">
+                <text-chunk font="helvetica" font-size="9">This is your ticket. Print this Entire page, fold it and bring it with you to the event.</text-chunk>
```
`helvetica` is the default font for text chunks. No need to specify the `font` attribute in this case.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,173 @@
+                    <division>
+                        <image src="qr-code" height="75" width="75" margin="15 0 0 10"></image>
+                        <paragraph text-align="top" margin="0 0 0 15">
+                            <text-chunk font="helvetica" font-size="10">{{.TicketNumber}}</text-chunk>
```
`helvetica` is the default font for text chunks. Also, `font-size` is 10 by default. No need to specify these attributes.

adrg reviewed (commented):
I left some observations, mostly regarding the template.

kellemNegasi commented:
Thanks @adrg  and @sampila for the input. I have addressed the reviews and updated the files. Can you check one more time if i have  missed some things.

sampila commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+            <text-chunk font="helvetica-bold" font-size="{{.FontSize}}">Epic Rock Concert</text-chunk>
+        </paragraph>
+        <paragraph text-align="left" margin="5 5 0 0">
+            <text-chunk font="helvetica" font-size="{{.SubFontSize}}">25.05.2021  7:30PM</text-chunk> 
```
Probably the date can be set to `Ticket` struct?

sampila reviewed (commented):
Overall looks good, added minor comment

sampila reviewed (approved):
Looks good to me

adrg commented on templates/concert-ticket/concert-ticket.json:
```diff
@@ -0,0 +1,44 @@
+            "field_value": "Coleman R. Merrell"
+        },
+        {
+            "field_name": "address",
```
Probably name this field `Address`, with a capital `A` so that it appears in the output as the other fields.

adrg commented on templates/concert-ticket/concert-ticket.json:
```diff
@@ -0,0 +1,44 @@
+        },
+        {
+            "field_name": "Name",
+            "field_value": "Coleman R. Merrell"
```
Can you change the name to `Jane R. Doe`, just to be on the safe side?
Also, don't forget to regenerate the screenshot.

adrg commented on templates/concert-ticket/pdf_concert_ticket.go:
```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Draw main content template.
```
Can you add an empty line above each of the comments in the `main` function please?

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+<table columns="2" margin="0 0 0 15">
+ <table-cell border-width-top="0.5" border-width-bottom="0.5" border-width-left="0.5" vertical-align="top" indent="0.5" border-style="single">
+    <division margin="0 0 15 15">
+        <paragraph margin="10 0 10 0">
```
Could use the shorthand form here for the margin attribute: `10 0` (top/bottom 10, left/right 0).

kellemNegasi replied:
I see.  
By the way could you point me to other usages of the tags and their attributes. I tried to get as much as I can from the source code of the parser but it would great if there is a documentation or something like that to refer to.

adrg replied:
I'm currently working on writing a tutorial/reference document regarding the usage of templates.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+ </table-cell>
+ <table-cell border-width-top="0.5" border-width-bottom="0.5" border-width-right="0.5" indent="0" border-style="single">
+    <division margin="0 0 15 15">
+        <paragraph text-align="left" margin="10 0 10 0">
```
Could use the shorthand form here for the margin attribute: 10 0 (top/bottom 10, left/right 0).
Also, please remove the `text-align` attribute. `left` is the default value already.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+        <paragraph>
+            <text-chunk font="helvetica-bold" font-size="{{.FontSize}}">Epic Rock Concert</text-chunk>
+        </paragraph>
+        <paragraph text-align="left" margin="5 5 0 0">
```
Please remove the `text-align` attribute. `left` is the default value already.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+            <table-cell>
+                    <division>
+                        <image src="qr-code" height="75" width="75" margin="15 0 0 10"></image>
+                        <paragraph text-align="top" margin="0 0 0 15">
```
`top` is not a valid value for the `text-align` attribute. Valid values are: `left`, `right`, `center` and `justify`. Please remove the attribute as you don't seem to be needing it anyway.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+                    <division margin="5 0 0 0">
+                        <table columns="2">
+                            {{range .Detail}}
+                            {{if ne .FieldName "address"}}
```
If you change the `address` field name to `Address`, probably need to update here as well.

adrg commented on templates/concert-ticket/templates/main.tpl:
```diff
@@ -0,0 +1,168 @@
+                <text-chunk font-size="9">A City literally built on Rock n' Roll in Cadott WI, </text-chunk>
+                <text-chunk font-size="9"> Rock Fest is the true Rock experience you can't miss. In its 26th year, </text-chunk>
+                <text-chunk font-size="9"> it is the top venue for people of all ages to come together for one common purpose: </text-chunk>
+                <text-chunk font-size="9">to congregate with other rock fans from across the world, in a place where rock music still matters.</text-chunk>
```
Please add a space after `matters.` so that there is a space between the sentences.

Please do that throughout the template so that the form is: 
`sentence A. sentence B.`
instead of
`sentence A.sentence B.`

adrg reviewed (commented):
Looks good overall. I left a couple more minor adjustments to make.

adrg reviewed (approved):
Looks good. Thank you for addressing the review points.
